File: Templates/LazySeg.java
class LazySeg {
    int l, r;
    long v = 0, lzAdd = 0, lzSet = 0;
    LazySeg left, right;
    public LazySeg(int l, int r, long[] a) {
        this.l = l; this.r = r;
        if (l == r) {
            v = a[l];
            return;
        }
        int mid = (l + r) / 2;
        left = new LazySeg(l, mid, a); right = new LazySeg(mid + 1, r, a);
        pull();
    }
    public LazySeg(long[] a) { this(0, a.length - 1, a); }
    public void push() {
        if (l == r) {
            if (lzSet != 0) v = lzSet;
            v += lzAdd;
            lzSet = 0; lzAdd = 0;
            return;
        }
        if (lzSet != 0) {
            v = (r - l + 1) * lzSet;
            left.lzSet = lzSet; left.lzAdd = 0;
            right.lzSet = lzSet; right.lzAdd = 0;
            lzSet = 0;
        }
        v += (r - l + 1) * lzAdd;
        left.lzAdd += lzAdd; right.lzAdd += lzAdd;
        lzAdd = 0;
    }
    public void pull() { v = left.v + right.v; }
    public long qry(int ll, int rr) {
        push();
        if (ll <= l && r <= rr) return v;
        if (r < ll || rr < l) return 0;
        return left.qry(ll, rr) + right.qry(ll, rr);
    }
    public void updAdd(int ll, int rr, long x) {
        if (ll <= l && r <= rr) {
            lzAdd += x;
            push();
            return;
        }
        push();
        if (r < ll || rr < l) return;
        left.updAdd(ll, rr, x); right.updAdd(ll, rr, x);
        pull();
    }
    public void updSet(int ll, int rr, long x) {
        if (ll <= l && r <= rr) {
            lzSet = x;
            lzAdd = 0;
            push();
            return;
        }
        push();
        if (r < ll || rr < l) return;
        left.updSet(ll, rr, x); right.updSet(ll, rr, x);
        pull();
    }
}

File: Templates/LazySeg.cpp
template<typename M> class LazySeg {
private:
    using T = M::T;
    using U = M::U;
    int n;
    V<T> tree;
    V<U> lazy;
    void pull(int i) {
        tree[i] = M::val_cmb(tree[2 * i], tree[2 * i + 1]);
    }
    void push(int i) {
        lazy[2 * i] = M::lazy_cmb(lazy[2 * i], lazy[i]);
        tree[2 * i] = M::apply_lazy(tree[2 * i], lazy[i]);
        lazy[2 * i + 1] = M::lazy_cmb(lazy[2 * i + 1], lazy[i]);
        tree[2 * i + 1] = M::apply_lazy(tree[2 * i + 1], lazy[i]);
        lazy[i] = M::lazy_id;
    }
    void upd(int i, int tl, int tr, int l, int r, U v) {
        if (r <= tl || tr <= l)
            return;
        if (l <= tl && tr <= r) {
            lazy[i] = M::lazy_cmb(lazy[i], v);
            tree[i] = M::apply_lazy(tree[i], v);
        } else {
            push(i);
            int tm = midpoint(tl, tr);
            upd(2 * i, tl, tm, l, r, v);
            upd(2 * i + 1, tm, tr, l, r, v);
            pull(i);
        }
    }
    T qry(int i, int tl, int tr, int l, int r) {
        if (r <= tl || tr <= l)
            return M::val_id;
        if (l <= tl && tr <= r)
            return tree[i];
        push(i);
        int tm = midpoint(tl, tr);
        return M::val_cmb(qry(2 * i, tl, tm, l, r), qry(2 * i + 1, tm, tr, l, r));
    }
public:
    LazySeg(int _n) {
        n = 1;
        while (n < _n)
            n *= 2;
        tree.assign(2 * n, M::val_id);
        lazy.assign(2 * n, M::lazy_id);
    }
    LazySeg(const V<T>& a) {
        int _n = ssize(a);
        n = 1;
        while (n < _n)
            n *= 2;
        tree.assign(2 * n, M::val_id);
        for (int i = 0; i < _n; i++)
            tree[i + n] = a[i];
        for (int i = n - 1; i >= 1; i--)
            pull(i);
        lazy.assign(2 * n, M::lazy_id);
    }
    void range_update(int l, int r, U v) {
        upd(1, 0, n, l, r, v);
    }
    T range_query(int l, int r) {
        return qry(1, 0, n, l, r);
    }
};

struct MaxAddPolicy {
    using T = ll; // value type
    using U = ll; // lazy type
    constexpr static T val_id = 0;
    constexpr static U lazy_id = 0;
    static T val_cmb(T a, T b) {
        return max(a, b);
    }
    static U lazy_cmb(U a, U b) { // later update second
        return a + b;
    }
    static T apply_lazy(T a, U b) {
        return a + b;
    }
};
